### PR TITLE
Changes reorder column events to use source and destination

### PR DESCRIPTION
**Pull request by hoonji**
## Motivation for features / changes
To prepare for showing shared hparam columns across runs and scalar tables, we need to unify the column re-ordering logic to use the same source/destination parameters (introduced in #6727).

## Technical description of changes
- Changes the DataTable component (shared by runs and scalar tables) to use the new ReorderColumnEvent type when column order is changed. The changes propagate to scalar card scalar column editor containers. Most of the changes are getting the tests to pass with the new signature.
- Changes the metrics reducers to correctly handle these new events. Logic is mostly unchanged, except that sort order will now be preserved even after column toggle (previously, toggling would group all disabled columns together, which destroys relative column order

misc:
- creates moveColumn utility to abstract shared logic between the hparams and metrics reducers. Removes redundant tests in hparams reducer

## Detailed steps to verify changes work correctly (as executed by you)
- Unit tests pass
- Manually tested sorting behavior

### CODE REVIEW

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_data_table.ng.html [hunk 1]
```diff
@@ -18,10 +18,8 @@
 <tb-data-table
   [headers]="columnHeaders"
   [sortingInfo]="sortingInfo"
-  [columnCustomizationEnabled]="columnCustomizationEnabled"
```
Comment (hoonji) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_data_table.ng.html:21]: unused

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_data_table.ng.html [hunk 2]
```diff
@@ -18,10 +18,8 @@
...
   (sortDataBy)="sortDataBy.emit($event)"
-  (orderColumns)="orderColumns($event)"
-  (removeColumn)="removeColumn.emit($event)"
```
Comment (hoonji) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_data_table.ng.html:24]: scalar tables shouldn't be able to remove columns, just hide them. But there will be no functional change atm because this is currently disabled.

File: tensorboard/webapp/metrics/views/right_pane/scalar_column_editor/scalar_column_editor_container.ts [hunk 3]
```diff
@@ -48,16 +56,27 @@ import {DataTableMode} from '../../../../widgets/data_table/types';
 export class ScalarColumnEditorContainer {
   constructor(private readonly store: Store<State>) {}
 
-  readonly singleHeaders$ = this.store.select(getSingleSelectionHeaders);
-  readonly rangeHeaders$ = this.store.select(getRangeSelectionHeaders);
+  readonly singleHeaders$ = this.store
```
Comment (hoonji) [inline tensorboard/webapp/metrics/views/right_pane/scalar_column_editor/scalar_column_editor_container.ts:59]: runs are now hidden in the edit menu so they will no longer be sortable or hidable

File: tensorboard/webapp/persistent_settings/_data_source/persistent_settings_data_source.ts [hunk 4]
```diff
@@ -235,7 +235,8 @@ export class OSSSettingsConverter extends SettingsConverter<
     if (
       Array.isArray(backendSettings.singleSelectionHeaders) &&
       // If the settings stored in the backend are invalid, reset back to default.
-      backendSettings.singleSelectionHeaders[0].name !== undefined
+      backendSettings.singleSelectionHeaders[0].name !== undefined &&
+      backendSettings.singleSelectionHeaders[0].type === 'RUN'
```
Comment (hoonji) [inline tensorboard/webapp/persistent_settings/_data_source/persistent_settings_data_source.ts:239]: adds check when loading from existing localstorage as downstream sorting logic depends on RUN being first

File: tensorboard/webapp/widgets/data_table/utils.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    }
+  }
+
+  const newColumns = [...columns];
+  newColumns.splice(sourceIndex, 1);
+  newColumns.splice(destinationIndex, 0, source);
+  return newColumns;
+}
+
+export const DataTableUtils = {
```
Comment (hoonji) [inline tensorboard/webapp/widgets/data_table/utils.ts:52]: this export style is ugly, but necessary due to https://github.com/tensorflow/tensorboard/pull/5991

File: tensorboard/webapp/widgets/data_table/data_table_component.ts [hunk 6]
```diff
@@ -181,13 +178,16 @@ export class DataTableComponent implements OnDestroy, AfterContentInit {
     if (!this.draggingHeaderName || !this.highlightedColumnName) {
       return;
     }
+    const source = this.getHeaderByName(this.draggingHeaderName);
+    const destination = this.getHeaderByName(this.highlightedColumnName);
+    if (source && destination && source !== destination) {
```
Comment (rileyajones) [inline tensorboard/webapp/widgets/data_table/data_table_component.ts:183]:
```suggestion
    const source = {...this.draggingHeaderName};
    const destination = {...this.highlightedColumnName};
    if (source && destination && source.name !== destination.name) {
```

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_data_table.ng.html (hunk 2 above)
Comment (hoonji) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_data_table.ng.html:24]: I agree that showing both "hide" and "remove" buttons is also reasonable, but showing both in the same context menu might make a lot of users wondering what the difference is. Meanwhile, allowing only one of these functionalities per table _might_ make the distinction a bit clearer.

In any case, I think it's important to gather more feedback from the team and from users on what the ideal context menu setup should be. It should be trivial to add/remove items later as necessary.

File: tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts [hunk 7]
```diff
@@ -701,6 +613,13 @@ describe('hparams/_redux/hparams_reducers_test', () => {
         })
       );
 
+      // Edge cases are tested by moveColumn tests.
```
[the remark this answers is not in the document]
Comment (hoonji) [reply tensorboard/webapp/hparams/_redux/hparams_reducers_test.ts:616]: I think we're on the same page here - the moveColumn spy is already in place :)

File: tensorboard/webapp/metrics/store/metrics_reducers.ts [hunk 8]
```diff
@@ -1434,34 +1431,30 @@ const reducer = createReducer(
...
-    // All enabled headers appear above all disabled headers.
-    headers.forEach((header) => {
-      if (header.enabled) {
-        enabledNewHeaders.push(header);
-      } else {
-        disabledNewHeaders.push(header);
+  on(
+    actions.dataTableColumnEdited,
+    (state, {source, destination, side, dataTableMode}) => {
+      let headers =
```
[the remark this answers is not in the document]
Comment (hoonji) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:1437]: Need `let` here as it's going to be reassigned in the next statement. I can also just create a new `const` reference if that's preferred.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts [hunk 9]
```diff
@@ -1434,34 +1431,30 @@ const reducer = createReducer(
...
+        dataTableMode === DataTableMode.RANGE
+          ? [...state.rangeSelectionHeaders]
```
[the remark this answers is not in the document]
Comment (hoonji) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:1439]: To clarify:
- `dataTableColumnEdited` is currently only used for reordering
- runs and hparams features already define separate actions for reordering and adding (e.g. `runsTableHeaderAdded`, `runsTableHeaderOrderChanged`)

To prevent confusion (and for consistency) I'll change `dataTableColumnEdited` -> `dataTableColumnOrderChanged`

[the remark this answers is not in the document]
Comment (hoonji) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:1439]: (a separate `dataTableColumnAdded` action will be added in an upcoming PR)

File: tensorboard/webapp/metrics/store/metrics_reducers.ts [hunk 10]
```diff
@@ -1471,40 +1464,24 @@ const reducer = createReducer(
...
-        currentToggledHeaderIndex,
-        newToggledHeaderIndex,
-        targetedHeaders
-      );
-
-      newHeaders[newToggledHeaderIndex] = {
-        ...newHeaders[newToggledHeaderIndex],
-        enabled: !newHeaders[newToggledHeaderIndex].enabled,
-      };
+      const newHeaders = targetedHeaders.map((header) => {
```
[the remark this answers is not in the document]
Comment (hoonji) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:1471]: The previous code essentially separates enabled and disabled headers into different groups, which I understand can be useful for quickly viewing all the enabled headers in one place. However, this feature had the drawback of destroying current sort order info, which will be especially confusing if we also add this behavior to hparam columns.

For example, assume the runs table has columns `| conv_layers | conv_kernel_size | dropout |`, then we hide `conv_kernel_size` from the scalar table. The old logic would also force-reorder the runs table columns to `| conv_layers | dropout |  conv_kernel_size |`, which is a very unexpected result of a "hide" operation.

In short, I think it'll be preferable to make "hide" just hide and not also influence sort order. Users can manually reorder the headers if they want to replicate the old behavior - I'm hoping it should be a minor inconvenience vs the large potential confusion caused by sudden sort changes.

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_container.ts [hunk 11]
```diff
@@ -686,15 +674,18 @@ export class ScalarCardContainer implements CardRenderer, OnInit, OnDestroy {
...
-    this.store.dispatch(dataTableColumnEdited(headerEditInfo));
+  editColumnHeaders({
+    source,
+    destination,
+    side,
+    dataTableMode,
+  }: HeaderEditInfo) {
+    this.store.dispatch(
+      dataTableColumnEdited({source, destination, side, dataTableMode})
+    );
```
[the remark this answers is not in the document]
Comment (hoonji) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_container.ts:685]: Oh these will need to be split up in a later PR, but I forgot to re-simplify in this one. Fixed!

File: tensorboard/webapp/metrics/views/right_pane/scalar_column_editor/scalar_column_editor_component.ts [hunk 12]
```diff
@@ -108,14 +99,20 @@ export class ScalarColumnEditorComponent implements OnDestroy {
...
-        getIndexOfColumn(this.draggingHeader, headers),
-        getIndexOfColumn(this.highlightedHeader, headers),
-        headers
-      ),
-    });
+    const source = this.getHeaderByName(headers, this.draggingHeader.name);
+    const destination = this.getHeaderByName(
+      headers,
+      this.highlightedHeader.name
+    );
```
[the remark this answers is not in the document]
Comment (hoonji) [reply tensorboard/webapp/metrics/views/right_pane/scalar_column_editor/scalar_column_editor_component.ts:106]: Agree this is better. Done!

File: tensorboard/webapp/metrics/views/right_pane/scalar_column_editor/scalar_column_editor_component.ts [hunk 13]
```diff
@@ -108,14 +99,20 @@ export class ScalarColumnEditorComponent implements OnDestroy {
...
+    if (source && destination && source !== destination) {
```
[the remark this answers is not in the document]
Comment (hoonji) [reply tensorboard/webapp/metrics/views/right_pane/scalar_column_editor/scalar_column_editor_component.ts:107]: Done

File: tensorboard/webapp/metrics/views/right_pane/scalar_column_editor/scalar_column_editor_container.ts [hunk 14]
```diff
@@ -48,16 +56,27 @@ import {DataTableMode} from '../../../../widgets/data_table/types';
...
-    this.store.dispatch(dataTableColumnEdited(editInfo));
+  onScalarTableColumnEdit({
+    source,
+    destination,
+    side,
+    dataTableMode,
+  }: HeaderEditInfo) {
+    this.store.dispatch(
+      dataTableColumnEdited({source, destination, side, dataTableMode})
+    );
```
[the remark this answers is not in the document]
Comment (hoonji) [reply tensorboard/webapp/metrics/views/right_pane/scalar_column_editor/scalar_column_editor_container.ts:79]: Same reason as earlier - it will be a downstream PR. Fixed!

File: tensorboard/webapp/metrics/views/right_pane/scalar_column_editor/scalar_column_editor_container.ts [hunk 15]
```diff
@@ -48,16 +56,27 @@ import {DataTableMode} from '../../../../widgets/data_table/types';
...
+    .pipe(map(headersWithoutRuns));
+  readonly rangeHeaders$ = this.store
+    .select(getRangeSelectionHeaders)
+    .pipe(map(headersWithoutRuns));
   readonly selectedTab$ = this.store.select(getTableEditorSelectedTab);
 
-  onScalarTableColumnToggled(toggleInfo: HeaderToggleInfo) {
-    this.store.dispatch(dataTableColumnToggled(toggleInfo));
+  onScalarTableColumnToggled({dataTableMode, header}: HeaderToggleInfo) {
+    this.store.dispatch(dataTableColumnToggled({dataTableMode, header}));
```
[the remark this answers is not in the document]
Comment (hoonji) [reply tensorboard/webapp/metrics/views/right_pane/scalar_column_editor/scalar_column_editor_container.ts:68]: Done

File: tensorboard/webapp/widgets/data_table/data_table_component.ts (hunk 6 above)
Comment (hoonji) [reply tensorboard/webapp/widgets/data_table/data_table_component.ts:183]: Doesn't work in this case because the instance variables are just names, not objects

Comment (rileyajones) [reply tensorboard/webapp/widgets/data_table/data_table_component.ts:183]: Ahhh, I missed that

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_data_table.ng.html (hunk 2 above)
Comment (rileyajones) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_data_table.ng.html:24]: We definitely shouldn't show both in the same context menu. I'm not sure splitting the functionality by table will be clearer for users.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:1439]: In that case can you `headers` from the props?

File: tensorboard/webapp/metrics/store/metrics_reducers.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:1471]: Fair enough, that does seem unexpected.

File: tensorboard/webapp/metrics/store/metrics_reducers.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (hoonji) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:1439]: Wait sorry - I need to correct myself. I _did_ add a `dataTableColumnAdded` in #6737, but it's unused because we don't actually have a feature for adding standard columns, only hparam columns. Thus, only the hparams action `dashboardHparamColumnAdded` will be used to add new columns.

Regarding passing header to the props, I propose that an action signature like [dashboardHparamColumnAdded](https://github.com/tensorflow/tensorboard/blob/4d7e337c9ad28cde11fbd8df981a9a235a27e2fc/tensorboard/webapp/hparams/_redux/hparams_reducers.ts#L93), i.e. only specifying {column, nextTo, side} and letting the reducer figure out how put it in the right place, is more in line with redux best practices, which advocate [putting logic in reducers, not components](https://redux.js.org/style-guide/#put-as-much-logic-as-possible-in-reducers); passing in headers via the action implies that the component must do some data manipulation first.

(Note that the previously implemented [runs table add action](https://github.com/tensorflow/tensorboard/blob/4d7e337c9ad28cde11fbd8df981a9a235a27e2fc/tensorboard/webapp/runs/store/runs_reducers.ts#L436) already uses the "put logic in reducers" pattern, albeit with a slightly different signature.)

[the remark this answers is not in the document]
Comment (rileyajones) [reply tensorboard/webapp/metrics/store/metrics_reducers.ts:1439]: I fully agree with everything in this statement.
The thing that I find confusing is that the action [dataTableColumnEdited](https://github.com/tensorflow/tensorboard/blob/4d7e337c9ad28cde11fbd8df981a9a235a27e2fc/tensorboard/webapp/metrics/actions/index.ts#L237) takes [HeaderEditInfo](https://github.com/tensorflow/tensorboard/blob/master/tensorboard/webapp/metrics/internal_types.ts#L97) as props which has an unused attribute `headers`

**Review by rileyajones — APPROVED** [review]
Thanks for sticking with this. I would really like to see the props changed on `dataTableColumnEdited` but otherwise things look pretty good.
go/small-cls

**Comment by hoonji** [comment]
> I would really like to see the props changed on dataTableColumnEdited

Ack - `headers` should be gone in the latest version

> go/small-cls

Yeah my bad, I didn't realize the code would get so bloated after tests 😭 I'll take care to make them smaller next time. Thanks btw for the comprehensive reviews on these huge batches of code!
